[#40][feature] 노드 생성 기능 - 노드 엔티티 구현

jihye-woo opened the pull request:
**Node 엔티티**와 노드 간 parent-child 관계를 나타내는 **ParentChild 엔티티**를 구현했습니다

> 관계 테이블과 매핑되는 **ParentChild 엔티티**의 네이밍은 변경가능합니다. 
> 추천하실만한 이름이 있다면 코멘트 남겨주세요!

-----

### 고민 되었던 부분 : parent와 child 연관관계
- Node 엔티티에서 parent 필드와 child 필드는 어떤 매핑을 해야 좋을지 고민했습니다
- 그리고 제가 내린 결론은 아래와 같아요
```java
public class Node { // ...
    // child는 하나의 parent를 가진다
    @OneToOne(mappedBy = "parent")
    private ParentChild parent;

    // parent는 여러 개의 child를 가진다
    @OneToMany(mappedBy = "child", cascade = CascadeType.ALL)
    private List<ParentChild> children = new ArrayList<>();
  ```


![image](https://user-images.githubusercontent.com/74452069/134646477-1bf6acd7-bff5-4b9d-96ed-b3a05cff5373.png)

### 고민 되었던 부분 : Enum 적용

> 참고 : https://fir-screen-46c.notion.site/Main-position-6aef5e2fa54040ce9cd6d8abc34eca43

- Enum(EnumType.STRING)을 적용하게 되면 DB 상에서는 varchar(255) 타입으로 관리가 됩니다(mysql에서는)
- 아무래도 우리 서비스에서는 노드의 생성이 많아질 것이기 때문에 **enum을 필수적으로 적용할 필요가 없고**, **option이 2가지밖에 없는 경우**라면 boolean 타입으로 관리하는 것이 더 효율적일 것 같아요!

- 이런 관점에서, enum을 사용하지 않아도 될 것 같은 부분이 아래 두 가지라고 생각했습니다
> 1. 현재 노드가 Main인지 여부(MAIN/NONMAIN)
>     - 메인과 메인이 아닌 노드 이외에 다른 옵션을 따로 추가할 예정이 아니라면 boolean으로 변경해도 좋을 것 같아요
> 2. 노드의 Position(UP/DOWN)
>     - 노드의 위치 정보가 up과 down 이외의 다른 옵션을 추가할 필요가 없다면 마찬가지로 boolean으로 변경해도 좋을 것 같습니다

- 그래서 현재 코드에서는 해당 부분을 boolean 타입으로 관리하고 있습니다! 만약 모두 괜찮다고 말씀해주시면 API 문서에도 반영하겠습니당


-----


#### 혹시 틀렸다고 생각하시거나 피드백 주실 부분이 있다면 언제든 환영입니다~~!!
#### 읽어주셔서 감사합니다아🙇‍♀️

jihye-woo committed:
feature : Node 엔티티 구현

> Children 엔티티 생성
  - Node 간 순환 참조 관계를 나타내는 공통 테이블과 매핑
  - 이름은 추후 변동 가능

issue : #40

jihye-woo committed:
refactor : Children 엔티티명 변경 -> ParentChild

> Parent와 Child 간의 관계 테이블(및 엔티티)임을 직관적으로 보여주기 위해 엔티티명 변경

issue : #40

jihye-woo committed:
feature : Node 엔티티 parent 필드의 연관관계 변경

> Node 엔티티의 parent와 ParentChild 연관관계 수정
  - 1:N ➡️ 1:1

issue : #40

jihye-woo committed:
feature : NodeRepository 생성

> 노드 엔티티 데이터 관리를 위한 Repository 생성

issue : #40

jihye-woo committed:
feature : Node 생성자 추가 및 팩토리 메소드 생성

> Node 팩토리 메소드 생성
  - createMainNode() : parent 노드를 받아오지 않아도 됨
  - createNonMainNode()

> Node private 생성자 구현
  - Node 생성에 반드시 필요한 필드만 파라미터로

> equals() hashCode() 메소드 재정의
  - 엔티티는 id만 같다면 동일한 엔티티로 취급

> index long 타입으로 변경
  - 100 단위로 증가하기 때문에 int보다 큰 단위인 long으로 변경

issue : #40

jihye-woo committed:
feature : ParentChild 생성자 추가 및 팩토리 메소드 생성

> ParentChild 팩토리 메소드 create() 구현

> private 생성자 구현
  - ParentChild 생성에 반드시 필요한 필드만 파라미터로

> equals() hashCode() 메소드 재정의
  - 엔티티는 id만 같다면 동일한 엔티티로 취급

issue : #40

jihye-woo committed:
test : 노드 생성 관련 테스트 코드 작성

> 메인 노드 생성 테스트 구현
> 자식 노드 생성 테스트 구현

issue : #40

sanhee commented on pintree/src/main/java/com/trio/pintree/login/domain/Node.java:
```diff
@@ -0,0 +1,85 @@
+    @Column(name = "node_id")
+    private Long id;
+
+    @Column(name = "is_main")
+    private boolean isMain;
+
+    @Column(name = "is_up")
+    private boolean isUp;
```
저는 boolean으로 관리하는 거 찬성입니다!

우디 더굽ㄴ에 enum string이 성능적인 이슈가 있다는 것을 처음 알게됐는데, 공유해주셔서 감사합니다 ㅎㅎ

pbg0205 replied:
저도 이 부분에 관해서 동의합니다!
다만 default value에 관해 고민해볼 필요가 있는 것 같아요. 이 필드를 선언한 목적은 해당 브랜치가 `main`에서 `non-main`으로 변경했을 때, 자동 배치를 위한 요구사항에 관련한 필드값으로 기억하고 있어요.
만약 처음부터 main으로 할당한 노드가 non-main으로 변경할 경우, `부모 브랜치의 위로 가야할까? 아래로 가야할까?`하는 의문점이 들었습니다. 이 부분은 회고 시간에 논의해보고 `default value`를 설정하면 좋을 것 같아요!

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/domain/Node.java:
```diff
@@ -0,0 +1,85 @@
+
+@Entity
+@Getter
+@Table(name = "node")
```
저희 생각해보니까 table naming 방식을 안정한거 같네요. 오늘 저녁 회고에 해당 내용에 관해 이야기하면 좋을거 같아요👍

sanhee commented on pintree/src/test/java/com/trio/pintree/login/domain/NodeTest.java:
```diff
@@ -0,0 +1,47 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DataJpaTest
+class NodeTest {
```
오 `@DataJpaTest` 처음 본 애노테이션이네요!..

제가 테스트를 잘 몰라서 질문합니다!

혹시 `@DataJpaTest`를 꼭 사용한 이유가 어떻게 되나요??
springboot 테스트나 mock 테스트 또는 
그냥 메모리db 레포지토리를 만들어서 테스트를 하는 등 여러 테스트 방법이 있는 것 같아서 

우디가 작성한 의도가 궁금해서 질문드렸습니다!

pbg0205 replied:
제가 알고 있기로는 `@DataJpaTest`를 사용하면 인메모리 기반으로 동작하는 것으로 알고 있어요!

jihye-woo replied:
넹! 맞아요! 쿠퍼 말씀대로 인메모리 기반으로 동작하고 

JPA 관련 테스트 설정만 로드하기 때문에 Repository 레벨에서 테스트를 진행할거라면 이게 훨씬 효율적일 것 같아서 사용해봤습니당!

[스택오버플로우 글](https://stackoverflow.com/a/59097848)에서 세가지 테스트 어노테이션의 각 특징에 대해 잘 설명해주셔서 참고하시면 좋을 것 같아요!

`@SpringBootTest` loads full application context
`@WebMvcTest` loads only the web layer, which includes security, filter, interceptors, etc for handling request/response
`@DataJpaTest` loads only configuration for JPA.

sanhee reviewed (commented):
고생하셨습니다 우디 ㅎㅎ

`partent_child <-> child node` 부분 관계가 조금 햇갈리는데, 이따 스크럼때 질문드리도록 할게요!
![image](https://user-images.githubusercontent.com/73760074/134659511-582c8937-09c5-428a-ada0-42965181a2c3.png)

그외 궁금한 질문 한개 드렸습니다 ㅎㅎ

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/domain/ParentChild.java:
```diff
@@ -9,10 +9,11 @@
-    @Id @GeneratedValue
-    @Column(name = "children_id")
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
저는 `AUTO`로 선언했는데 `IDENTITY`을 선언하는 방식이 더 좋은 것 같습니다.
테스트 해보니 DB를 덜 조회하는데 혹시 이유에 대해 아시면 말씀 부탁드려요 👍 👍

jihye-woo replied:
DB를 덜 조회한다는 말을 제대로 이해하지 못했네요ㅠㅠ 혹시 좀 더 자세히 설명해주실 수 있나요?

pbg0205 replied:
이 부분이 아까 실행해본 코드였는데 `call next value for hibernate_sequence` 이 부분도 제가 알기로는 PK값을 받아오기 위해 DB에 갔다오는걸로 알고 있어요. AUTO와 IDENTITY를 호출할 때, query 순서가 조금 다르지만 DB를 갔다오는 횟수는 `IDENTITY`가 적다고 생각했어요😄 이 부분은 저도 정확히 알고 있는 부분이 아니라서 추가적인 학습이 필요할거 같아요ㅠ
```java
public class Main {

    public static void main(String[] args) {
        EntityManagerFactory entityManagerFactory = Persistence.createEntityManagerFactory("hello");
        EntityManager entityManager = entityManagerFactory.createEntityManager();
        EntityTransaction transaction = entityManager.getTransaction();

        transaction.begin();
        try {

            System.out.println("> teamA 추가");
            Team team1 = new Team();
            team1.setName("teamA");
            entityManager.persist(team1);
            System.out.println("======================");

            System.out.println("> teamB 추가");
            Team team2 = new Team();
            team2.setName("teamB");
            entityManager.persist(team2);
            System.out.println("======================");

            System.out.println("> member1 추가");
            Member member1 = new Member();
            member1.setName("memberA");
            entityManager.persist(member1);
            member1.setTeam(team1);
            System.out.println("======================");

            System.out.println("> member2 추가");
            Member member2 = new Member();
            member2.setName("memberB");
            entityManager.persist(member2);
            member2.setTeam(team2);
            System.out.println("======================");

            transaction.commit();
        } catch (Exception e) {
            transaction.rollback();
            e.printStackTrace();
        } finally {
            entityManager.close();
        }
    }
}
```

<br>

`AUTO`
```
>>> teamA 추가
Hibernate: 
    call next value for hibernate_sequence
======================
>>> teamB 추가
Hibernate: 
    call next value for hibernate_sequence
======================
>>> member1 추가
Hibernate: 
    call next value for hibernate_sequence
======================
>>> member2 추가
Hibernate: 
    call next value for hibernate_sequence
======================
>>> flush
Hibernate: 
    /* insert jpa.basic.domain.Team
        */ insert 
        into
            Team
            (name, id) 
        values
            (?, ?)
Hibernate: 
    /* insert jpa.basic.domain.Team
        */ insert 
        into
            Team
            (name, id) 
        values
            (?, ?)
Hibernate: 
    /* insert jpa.basic.domain.Member
        */ insert 
        into
            Member
            (name, TEAM_ID, MEMBER_ID) 
        values
            (?, ?, ?)
Hibernate: 
    /* insert jpa.basic.domain.Member
        */ insert 
        into
            Member
            (name, TEAM_ID, MEMBER_ID) 
        values
            (?, ?, ?)
Hibernate: 
    /* update
        jpa.basic.domain.Member */ update
            Member 
        set
            name=?,
            TEAM_ID=? 
        where
            MEMBER_ID=?
Hibernate: 
    /* update
        jpa.basic.domain.Member */ update
            Member 
        set
            name=?,
            TEAM_ID=? 
        where
            MEMBER_ID=?
======================
```
<br>

`IDENTITY`
```
>>>  teamA 추가
Hibernate: 
    call next value for hibernate_sequence
======================
>>>  teamB 추가
Hibernate: 
    call next value for hibernate_sequence
======================
>>>  member1 추가
Hibernate: 
    /* insert jpa.basic.domain.Team
        */ insert 
        into
            Team
            (name, id) 
        values
            (?, ?)
Hibernate: 
    /* insert jpa.basic.domain.Team
        */ insert 
        into
            Team
            (name, id) 
        values
            (?, ?)
Hibernate: 
    /* insert jpa.basic.domain.Member
        */ insert 
        into
            Member
            (MEMBER_ID, name, TEAM_ID) 
        values
            (null, ?, ?)
======================
>>> member2 추가
Hibernate: 
    /* insert jpa.basic.domain.Member
        */ insert 
        into
            Member
            (MEMBER_ID, name, TEAM_ID) 
        values
            (null, ?, ?)
======================
>>>flush
Hibernate: 
    /* update
        jpa.basic.domain.Member */ update
            Member 
        set
            name=?,
            TEAM_ID=? 
        where
            MEMBER_ID=?
Hibernate: 
    /* update
        jpa.basic.domain.Member */ update
            Member 
        set
            name=?,
            TEAM_ID=? 
        where
            MEMBER_ID=?
======================
```

jihye-woo replied:
호옹 싱기하군요 이런 차이가 있는지는 몰랐어요! 자세하게 설명해주셔서 감사합니다!!
제가 Identity로 설정한 이유는 아무래도 AUTO 설정이 database dialect에 맞는 '적절한' 생성 전략을 찾아주는 기능이다보니 예상치 못한 상황이 발생할 수도 있지 않을까라는 우려에서였어요!

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/domain/ParentChild.java:
```diff
@@ -9,10 +9,11 @@
-    @Column(name = "children_id")
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "parent_child_id")
```
column name 명시하는 방식 좋네요! 저도 코드짤 때 적용해 봐야겠어요😄

jihye-woo replied:
확실히 코드 작업하다가 필드명이 바뀌는바람에 생기는 문제가 적어질 것 같아서 좋을 것 같더라구요!

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/domain/Node.java:
```diff
@@ -39,9 +41,45 @@
+        return node;
+    }
+
+    private void addChild(ParentChild parentChild) {
+        children.add(parentChild);
+    }
```
비록 ParentChild가 중간 테이블이지만 application에서의 데이터 일관성을 위해 `연관관계 편의 메서드`로 작성해보는 것은 어떨까요??😀

jihye-woo replied:
아래 두 메소드를 하나로 통합하자는 말씀으로 이해했는데, 이게 맞다면 전 좋습니다!
확실히 기능구현에만 초점을 맞추었는데, 이렇게 PR 리뷰를 받으면서 리펙토링할 지점들을 다시 한 번 점검해볼 수 있어서 좋네요ㅎㅎ 감사합니다😃
```java
private void addChild(ParentChild parentChild) {
    children.add(parentChild);
}

private void addParent(ParentChild parentChild) {
    this.parent = parentChild;
}
```

pbg0205 replied:
```java
private void addChild(ParentChild parentChild) {
    children.add(parentChild);
    parentChild.setChild(this); // 저는 이 부분 말씀드리는거였어요!
}
```

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/domain/Node.java:
```diff
@@ -39,9 +41,45 @@
 
-    @OneToMany(mappedBy = "child")
-    private List<Children> children = new ArrayList<>();
+    public static Node createMainNode(boolean isUp, long index, boolean isOfficial) {
+        return new Node(false, isUp, index, isOfficial);
+    }
```
이 부분에 default value로 `isUp = false`로 선언해놓으셨군요!

sanhee replied:
Node 클래스 빌더 패턴 사용하게 되면, 좀더 가독성이 좋아질 것 같아요!

jihye-woo replied:
@sanhee 확실히 필드들이 많아지다보니 빌더로 관리하는게 더 좋을 것 같네요!ㅎㅎ

@pbg0205 그래도 필드레벨에서 디폴트 값을 확인하면 좋을 것 같아서 쿠퍼 말씀대로 필드에 명시해놓는게 좋을 것 같아요!

모두 좋은 코멘트 감사합니다😃

jihye-woo replied:
빌더로 바꾸려하니 createNonMainNode() 같은 경우에는 Node를 생성하는 과정 뿐만 아니라 parent와 child의 관계를 연결해주는 로직도 함께 추가가 되어야해서 결국 메소드로 묶어야 하더라구요ㅠㅠ 일단 빌더는 다음 리펙토링때 다시 고민해보는 게 좋을 것 같아요!

```java
    public static Node createNonMainNode(Node parent, boolean isUp, long index, boolean isOfficial) {
        Node node = new Node(false, isUp, index, isOfficial); // 이 부분을 빌더로 변경하더라도 아래 과정과 함께 묶여야해서 결국 메소드를 호출해야하네요ㅠㅠ
        ParentChild parentChild = ParentChild.create(parent, node);
        parent.addChild(parentChild);
        node.addParent(parentChild);
        return node;
    }
```

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/domain/Node.java:
```diff
@@ -39,9 +41,45 @@
     @OneToMany(mappedBy = "child", cascade = CascadeType.ALL)
     private List<ParentChild> children = new ArrayList<>();
```
cascade 방식에 관해 한번 고민해볼 필요가 있는거 같아요.
제가 테스트해보았을 때는 자식이 부모 관계를 참조하고 있지 않을 경우, delete query가 발생하더라구요.
이 부분은 당장은 아니지만 삭제 로직을 구현하실 때, 참고해보시면 좋을 것 같아 코멘트 남겨놓아요! 😃
저도 한번 더 확인해보고 공유할 부분 있으면 말씀드리겠습니다ㅎㅎ

jihye-woo replied:
children 리스트에서 삭제 작업이 진행될 때 다시 한 번 보는게 좋겠지만 말씀하신대로 이 부분을 잘 염두에 두고 코드 작업을 진행하는 것이 좋겠네요
삭제 작업이 진행되면 ParentChild도 함께 삭제되어야하니까 cascade 뿐만 아니라 `orphanremoval = true` 옵션을 함께 사용하는 것도 좋을 것 같습니다!

pbg0205 commented on pintree/src/test/java/com/trio/pintree/login/domain/NodeTest.java:
```diff
@@ -0,0 +1,47 @@
+        repository.save(parent);
+
+        Node savedNode = repository.getById(parent.getId());
+        assertThat(savedNode.getChildren()).hasSize(2);
```
테스트로 size 확인하는 메서드가 존재하네요! 배워갑니다👍

pbg0205 replied:
`given`, `when`, `then` 빈번하게 사용할 경우, live template로 등록하는 방법이 있는데 필요하실까 해서 정리한 링크 첨부합니다ㅎㅎ
https://dune-store-e37.notion.site/TDD-live-template-80aa471056cd4a4f8b4ee8409bf2371b

pbg0205 reviewed (approved):
전반적으로 깔끔하게 작성하셔서 오히려 배워가네요 👍 👍
코드 작성하실 때, 혹시나 발생할 수 있는 부분에 관해서 간단한 코멘트도 첨부했어요.
한번 확인해보시고 참고하실 부분있으시면 참고하셨으면 좋겠습니다!
수고 많으셨어요 🤨 👍

jihye-woo committed:
fix : Node의 parent, children 필드의 mappedBy 대상 변경

> Node의 parent 필드
  - mappedBy ParentChild의 child

> Node의 children 필드
  - mappedBy ParentChild의 parent

issue : #43

jihye-woo committed:
refactor : Node의 boolean 타입 필드, default value 선언

> 필드 선언 시, default 값이 되는 변수도 함께 선언

issue : #43

jihye-woo commented:
일단 말씀해주신 부분 중 빌더와 연관관계 편의 메소드는 조금 더 고민을 해본뒤 이후의 리펙토링 과정에서 진행을 해보는 것이 좋을 것 같습니다!

그리고 노을이 의문을 가지셨던(?) 부분을 다시 생각해보고서 관계를 살짝 수정해보았습니다
<img width="585" alt="Screen Shot 2021-09-25 at 1 34 46 PM" src="https://user-images.githubusercontent.com/74452069/134758199-8744e6be-db45-4321-bfdb-c5374b87810c.png">

네이밍 때문에 헷갈렸네요ㅠㅠ 위 내용을 바탕으로 현재 수정된 코드는 아래와 같습니다
```java
    @OneToOne(mappedBy = "child") // 👈 👈 parent는 관계 테이블에서 child를 바라보고
    private ParentChild parent;

    @OneToMany(mappedBy = "parent", cascade = CascadeType.ALL) // 👈 👈 반대로 child는 관계 테이블에서 parent를 바라보고
    private List<ParentChild> children = new ArrayList<>();
```
